File: app/ui/src/api/catalog.ts
import type {
  CatalogExperimentDetail,
  CatalogExperimentListItem,
  CatalogList,
  CatalogMaterialModel,
  CatalogMaterialParameter,
  CatalogMaterialParameterDetail,
  CatalogMeta,
  CatalogQuantityKind,
  CatalogQuantityKindDetail,
  CatalogRuntimeSlice,
  CatalogRuntimeSliceRequest,
  CatalogSearchItem,
  CatalogSolverDetail,
  CatalogSolverListItem,
  ListQuery,
} from '@/contracts/catalog'
import {
  parseCatalogExperimentDetail,
  parseCatalogExperimentList,
  parseCatalogMaterialModel,
  parseCatalogMaterialModelList,
  parseCatalogMaterialParameterDetail,
  parseCatalogMaterialParameterList,
  parseCatalogMeta,
  parseCatalogQuantityKindDetail,
  parseCatalogQuantityKindList,
  parseCatalogRuntimeSlice,
  parseCatalogSearchResponse,
  parseCatalogSolverDetail,
  parseCatalogSolverList,
} from '@/contracts/catalogValidators'
import { request, type RequestContext } from './http'

export type CatalogExperimentIdentity = Readonly<{
  key: string
  namespace: string
  repository: string
  version: string
  coordinate: string
}>

function catalogUrl(path: string, query: ListQuery = {}) {
  const params = new URLSearchParams()
  Object.entries(query).forEach(([key, value]) => {
    if (value !== undefined && value !== '') params.set(key, String(value))
  })
  const encoded = params.toString()
  return `/catalog${path}${encoded ? `?${encoded}` : ''}`
}

export type {
  CatalogMaterialModel,
  CatalogMaterialParameter,
  CatalogMaterialParameterDetail,
  CatalogMeta,
  CatalogExperimentDetail,
  CatalogExperimentListItem,
  CatalogQuantityKind,
  CatalogQuantityKindDetail,
  CatalogRuntimeSlice,
  CatalogRuntimeSliceRequest,
  CatalogSearchItem,
  CatalogSolverDetail,
  CatalogSolverListItem,
} from '@/contracts/catalog'

export const catalogQueryKeys = {
  root: ['catalog'] as const,
  meta: ['catalog', 'meta'] as const,
  quantityKindsList: (query: ListQuery) => ['catalog', 'quantity-kinds', 'list', query] as const,
  quantityKindsInfinite: (query: ListQuery) => ['catalog', 'quantity-kinds', 'infinite', query] as const,
  quantityKind: (name: string) => ['catalog', 'quantity-kind', name] as const,
  materialParametersList: (query: ListQuery) => ['catalog', 'material-parameters', 'list', query] as const,
  materialParametersInfinite: (query: ListQuery) => ['catalog', 'material-parameters', 'infinite', query] as const,
  materialParameter: (key: string) => ['catalog', 'material-parameter', key] as const,
  materialModels: (query: ListQuery) => ['catalog', 'material-models', query] as const,
  materialModel: (key: string) => ['catalog', 'material-model', key] as const,
  materialManagerRuntime: (parameterNames: readonly string[]) =>
    ['catalog', 'material-manager-runtime', parameterNames] as const,
  recordedDataRuntime: (quantityKindNames: readonly string[]) =>
    ['catalog', 'recorded-data-runtime', quantityKindNames] as const,
  solvers: (query: ListQuery) => ['catalog', 'solvers', query] as const,
  solver: (name: string, version: string) => ['catalog', 'solver', name, version] as const,
  experiments: (query: ListQuery) => ['catalog', 'experiments', query] as const,
  experiment: (identity: string | CatalogExperimentIdentity) =>
    ['catalog', 'experiment', typeof identity === 'string' ? identity : identity.coordinate] as const,
  search: (query: string, limit: number) => ['catalog', 'search', query, limit] as const,
} as const

export const catalogApi = {
  meta: (context?: RequestContext) =>
    request<CatalogMeta>('get', catalogUrl('/meta'), undefined, {
      signal: context?.signal,
      validate: parseCatalogMeta,
    }),
  listQuantityKinds: (query: ListQuery = {}, context?: RequestContext) =>
    request<CatalogList<CatalogQuantityKind>>('get', catalogUrl('/quantity-kinds', query), undefined, {
      signal: context?.signal,
      validate: parseCatalogQuantityKindList,
    }),
  getQuantityKind: (name: string, context?: RequestContext) =>
    request<CatalogQuantityKindDetail>(
      'get',
      catalogUrl(`/quantity-kinds/${encodeURIComponent(name)}`),
      undefined,
      { signal: context?.signal, validate: parseCatalogQuantityKindDetail },
    ),
  listMaterialParameters: (query: ListQuery = {}, context?: RequestContext) =>
    request<CatalogList<CatalogMaterialParameter>>('get', catalogUrl('/material-parameters', query), undefined, {
      signal: context?.signal,
      validate: parseCatalogMaterialParameterList,
    }),
  getMaterialParameter: (key: string, context?: RequestContext) =>
    request<CatalogMaterialParameterDetail>(
      'get',
      catalogUrl(`/material-parameters/${encodeURIComponent(key)}`),
      undefined,
      { signal: context?.signal, validate: parseCatalogMaterialParameterDetail },
    ),
  listMaterialModels: (query: ListQuery = {}, context?: RequestContext) =>
    request<CatalogList<CatalogMaterialModel>>('get', catalogUrl('/material-models', query), undefined, {
      signal: context?.signal,
      validate: parseCatalogMaterialModelList,
    }),
  getMaterialModel: (key: string, context?: RequestContext) =>
    request<CatalogMaterialModel>(
      'get',
      catalogUrl(`/material-models/${encodeURIComponent(key)}`),
      undefined,
      { signal: context?.signal, validate: parseCatalogMaterialModel },
    ),
  listSolvers: (query: ListQuery = {}, context?: RequestContext) =>
    request<CatalogList<CatalogSolverListItem>>('get', catalogUrl('/solvers', query), undefined, {
      signal: context?.signal,
      validate: parseCatalogSolverList,
    }),
  getSolver: (name: string, version: string, context?: RequestContext) =>
    request<CatalogSolverDetail>(
      'get',
      catalogUrl(`/solvers/${encodeURIComponent(name)}/${encodeURIComponent(version)}`),
      undefined,
      { signal: context?.signal, validate: parseCatalogSolverDetail },
    ),
  listExperiments: (query: ListQuery = {}, context?: RequestContext) =>
    request<CatalogList<CatalogExperimentListItem>>('get', catalogUrl('/experiments', query), undefined, {
      signal: context?.signal,
      validate: parseCatalogExperimentList,
    }),
  async getExperiment(identity: string | CatalogExperimentIdentity, context?: RequestContext) {
    const parsed =
      typeof identity === 'string' && identity.startsWith('caemble:experiment/')
        ? identity.slice('caemble:experiment/'.length).split('/')
        : null
    const coordinateTail = parsed?.[parsed.length - 1]?.split('@')
    const key = typeof identity === 'string' ? (coordinateTail?.[0] ?? identity) : identity.key
    const query =
      typeof identity === 'string'
        ? parsed && coordinateTail
          ? { namespace: parsed[0], repository: parsed[1], version: coordinateTail[1] }
          : {}
        : { namespace: identity.namespace, repository: identity.repository, version: identity.version }
    return request<CatalogExperimentDetail>(
      'get',
      catalogUrl(`/experiments/${encodeURIComponent(key)}`, query),
      undefined,
      { signal: context?.signal, validate: parseCatalogExperimentDetail },
    )
  },
  search: async (q: string, limit = 50, context?: RequestContext) =>
    (
      await request<Readonly<{ items: readonly CatalogSearchItem[] }>>(
        'get',
        catalogUrl('/search', { q, limit }),
        undefined,
        { signal: context?.signal, validate: parseCatalogSearchResponse },
      )
    ).items,
  runtimeSlice: (payload: CatalogRuntimeSliceRequest, context?: RequestContext) =>
    request<CatalogRuntimeSlice>('post', catalogUrl('/runtime-slice'), payload, {
      signal: context?.signal,
      validate: parseCatalogRuntimeSlice,
    }),
} as const

File: app/api/app/service/catalog.py
from __future__ import annotations

import base64
from typing import Any

from caemble_catalog import Catalog

from catalog_models import CatalogRuntimeSliceRequest


def _offset(cursor: str | None) -> int:
    if cursor is None:
        return 0
    value = base64.urlsafe_b64decode(cursor.encode("ascii") + b"===").decode("ascii")
    return int(value)


def _cursor(offset: int, count: int, total: int) -> str | None:
    next_offset = offset + count
    if next_offset >= total:
        return None
    return (
        base64.urlsafe_b64encode(str(next_offset).encode("ascii"))
        .decode("ascii")
        .rstrip("=")
    )


def _revision(catalog: Catalog) -> str:
    return catalog.meta()["catalogRevision"]


def _page(
    catalog: Catalog,
    items: list[Any],
    total: int,
    offset: int,
) -> tuple[dict[str, Any], str]:
    return (
        {
            "items": items,
            "nextCursor": _cursor(offset, len(items), total),
            "total": total,
        },
        _revision(catalog),
    )


def catalog_meta(catalog: Catalog) -> tuple[dict[str, Any], str]:
    result = catalog.meta()
    return result, result["catalogRevision"]


def list_quantity_kinds(
    catalog: Catalog,
    *,
    query: str | None,
    domain: str | None,
    solver_name: str | None,
    solver_version: str | None,
    usage: str | None,
    unit: str | None,
    tensor_order: int | None,
    limit: int,
    cursor: str | None,
) -> tuple[dict[str, Any], str]:
    offset = _offset(cursor)
    items, total = catalog.list_quantity_kinds(
        query=query,
        domain=domain,
        solver_name=solver_name,
        solver_version=solver_version,
        usage=usage,
        unit=unit,
        tensor_order=tensor_order,
        limit=limit,
        offset=offset,
    )
    return _page(catalog, items, total, offset)


def get_quantity_kind(catalog: Catalog, name: str) -> tuple[dict[str, Any], str]:
    result = {**catalog.quantity_kind(name), **catalog.quantity_kind_relations(name)}
    return result, _revision(catalog)


def list_material_parameters(
    catalog: Catalog,
    *,
    query: str | None,
    domain: str | None,
    quantity_kind: str | None,
    solver_name: str | None,
    solver_version: str | None,
    limit: int,
    cursor: str | None,
) -> tuple[dict[str, Any], str]:
    offset = _offset(cursor)
    items, total = catalog.list_material_parameters(
        query=query,
        domain=domain,
        solver_name=solver_name,
        solver_version=solver_version,
        quantity_kind=quantity_kind,
        limit=limit,
        offset=offset,
    )
    return _page(catalog, items, total, offset)


def get_material_parameter(catalog: Catalog, key: str) -> tuple[dict[str, Any], str]:
    result = {
        **catalog.material_parameter(key),
        **catalog.material_parameter_relations(key),
    }
    return result, _revision(catalog)


def list_material_models(
    catalog: Catalog,
    *,
    query: str | None,
    limit: int,
    cursor: str | None,
) -> tuple[dict[str, Any], str]:
    offset = _offset(cursor)
    items, total = catalog.list_material_models(
        query=query,
        limit=limit,
        offset=offset,
    )
    return _page(catalog, items, total, offset)


def get_material_model(catalog: Catalog, key: str) -> tuple[dict[str, Any], str]:
    return catalog.material_model(key), _revision(catalog)


def list_solvers(
    catalog: Catalog,
    *,
    query: str | None,
    limit: int,
    cursor: str | None,
) -> tuple[dict[str, Any], str]:
    offset = _offset(cursor)
    items, total = catalog.page_solvers(query=query, limit=limit, offset=offset)
    return _page(catalog, items, total, offset)


def get_solver(
    catalog: Catalog,
    name: str,
    version: str,
) -> tuple[dict[str, Any], str]:
    return catalog.solver_detail(name, version), _revision(catalog)


def list_experiments(
    catalog: Catalog,
    *,
    query: str | None,
    solver_name: str | None,
    solver_version: str | None,
    namespace: str | None,
    repository: str | None,
    limit: int,
    cursor: str | None,
) -> tuple[dict[str, Any], str]:
    offset = _offset(cursor)
    items, total = catalog.list_experiments(
        query=query,
        solver_name=solver_name,
        solver_version=solver_version,
        namespace=namespace,
        repository=repository,
        limit=limit,
        offset=offset,
    )
    return _page(catalog, items, total, offset)


def get_experiment(
    catalog: Catalog,
    key: str,
    *,
    namespace: str | None,
    repository: str | None,
    version: str | None,
) -> tuple[dict[str, Any], str]:
    result = catalog.experiment(
        key,
        namespace=namespace,
        repository=repository,
        version=version,
    )
    return result, _revision(catalog)


def search_catalog(
    catalog: Catalog,
    query: str,
    *,
    limit: int,
) -> tuple[dict[str, Any], str]:
    return {"items": catalog.search(query, limit=limit)}, _revision(catalog)


def build_runtime_slice(
    catalog: Catalog,
    request: CatalogRuntimeSliceRequest,
) -> tuple[dict[str, Any], str]:
    result = catalog.runtime_slice(
        solvers=[(item.name, item.version) for item in request.solvers],
        quantity_kinds=request.quantityKinds,
        material_parameters=request.materialParameters,
        material_models=request.materialModels,
    )
    return result, _revision(catalog)
